logstash-core/src/main/java/org/logstash/plugin/Input.java

```diff
@@ -0,0 +1,32 @@
+package org.logstash.plugin;
+
+import org.logstash.Event;
+
+import java.util.Collection;
+import java.util.function.Consumer;
+
+/**
+ * A Logstash input plugin.
+ * <p>
+ * Input plugins produce events intended given to Processors.
+ * <p>
+ * Inputs will generally run forever until there is some terminating condition such as a pipeline shutdown.
+ */
+public interface Input {
+    /**
+     * The method contract is as follows:
+     * <p>
+     * 1) When `Thread.interrupted()` is true, you MUST close all open resources and return.
+     * 2) When this method returns, it is assumed all work for this input is completed.
+     * 3) consumer.accept(...) may block
+     * <p>
+     * Acknowledging Data:
+     * <p>
+     * When `consumer.accept(...)` returns, the given Events have been successfully written into the consumer, and it is now
+     * safe to acknowledge these events to the upstream data source. The consumer is generally the Logstash internal queue,
+     * such as the persistent queue.
+     *
+     * @param consumer Send batches of events with consumer.accept(batch).
+     */
+    void run(Consumer<Collection<Event>> consumer);
```

Comment:
I think a stronger contract here would be to sub the Consumer interface to satisfy the generic (or just duplicate the Consumer interface with the generic satisfied).  It will still be strong typed, imo slightly simpler to read.  Also, if you duplicate, we can change the semantics if `accept` isn't a good fit.

Reply:
Being explict about the type (`Collection<Event>`) feels like the right, rather than having:

```
public interface Input<T> {
  void run(Consumer<T> consumer);

  // or this
  void run(Consumer<Collection<T>> consumer);
}
```

because for Logstash inputs, I can't imagine any other type being passed. What did you have in mind?

Reply:
@jordansissel as discussed I think we should def not use `Collection` here.

What I really like would be this:

```java
public interface EventIterator {

    /**
     * @return Event currently pointed at
     */
    Event current();

    /**
     * @return True iff there was another Event in the iterator
     */
    boolean next();

    /**
     * @param buffer Collection to store all events in
     * @param count Maximum number of events to store in buffer
     * @return Number of events copied to buffer
     */
    int collect(Collection<Event> buffer, int count);
}
```

If we pass something like this down the callstack, we can hide the whole Queue logic behind it and each `Filter` and `Output` can decide if it wants to operate in a batched or non-batched way and we can optimize accordingly. Also we can simply signal end of execution by returning `false` from `next()` as well as `0` from `collect` (I think both can be made blocking, the current situation of returning empty batches under pressure is just wrong imo).

Reply:
I meant this:
```
public interface InputConsumer extends Consumer<Collection<Event>> {
}
```
so the API looks like
```
public interface Input {
    void run(InputConsumer consumer);
}
```
...or if you want better/different semantics then `apply` and `andThen` just create new functional interface and change the method names  (there is nothing magic about the Consumer interface).

It is a minor point, and just trying to simplify the readability.